src/runtime/dict.cpp

```diff
@@ -596,21 +596,33 @@ extern "C" Box* dictNew(Box* _cls, BoxedTuple* args, BoxedDict* kwargs) {
     return new (cls) BoxedDict();
 }
 
-void dictMerge(BoxedDict* self, Box* other) {
+bool dictMergeImpl(BoxedDict* self, Box* other) {
     if (PyDict_Check(other)) {
         for (const auto& p : static_cast<BoxedDict*>(other)->d)
             self->d[p.first] = p.second;
-        return;
+        return true;
     }
 
     static BoxedString* keys_str = internStringImmortal("keys");
     CallattrFlags callattr_flags{.cls_only = false, .null_on_nonexistent = true, .argspec = ArgPassSpec(0) };
     Box* keys = callattr(other, keys_str, callattr_flags, NULL, NULL, NULL, NULL, NULL);
-    assert(keys);
+    if (!keys) {
```

Comment:
ok sorry that's not quite right -- in the kwargs case (`ext_do_call` in ceval.c) they call PyDict_Update which calls PyDict_Merge directly.  The dict.update method (and `__init__`) calls `dict_update_common`, which tries either PyDict_Merge or PyDict_MergeFromSeq2 depending on whether the "keys" attribute exists.  We actually have this distinction I guess, so the assert should have just been an exception.  sorry about that.